### PR TITLE
Support for continuation of aborted training

### PR DESCRIPTION
**Pull request by davvil**
Transparently continue a training if it was aborted. **Important:** Start sockeye with exactly the same arguments.

Because of the asynchronous decoding for BLEU evaluation, some BLEU scores may get lost in the continuation, but this should not have a big effect on the final result.

Results are not yet 100% equal in the continuation, but are certainly in the same range. There must be still some part of the "traning state" missing, put it is surely usable, especially if dealing with long trainings.

### CODE REVIEW

File: sockeye/callback.py [hunk 1]
```diff
@@ -243,6 +244,21 @@ def stop_fit_callback(self):
         self._empty_decoder_metric_queue()
         self._write_scores()
 
+    def save_state(self, fname):
+        """
+        Save the state: current metrics and best checkpoint
```
Comment (fhieber) [inline sockeye/callback.py]: add
:param fname: Name of file to save the state to.

Also for the method below

Comment (fhieber) [inline sockeye/callback.py]: Change to SaveS the state [...]
Missing period at the end of the docstring.

File: sockeye/constants.py [hunk 2]
```diff
@@ -80,6 +80,17 @@
 METRICS_NAME = "metrics"
 TENSORBOARD_NAME = "tensorboard"
```
Comment (fhieber) [inline sockeye/constants.py]: Maybe add a comment:

```# training resumption constants```

File: sockeye/data_io.py [hunk 3]
```diff
@@ -426,13 +428,24 @@ def reset(self):
...
-            self.nd_label.append(mx.nd.array(self.data_label[i].take(indices, axis=0), dtype=self.dtype))
+            self.indices.append(np.random.permutation(len(self.data_source[i])))
+            self._append_ndarrays(i, self.indices[-1])
+
+    def _append_ndarrays(self, bucket: int, shuffled_indices: np.array):
+        """
+        Appends the actual data, selected by the given indices, to the NDArrays
+        of the appropriate bucket. Use when reshuffling the data.
+
+        :param bucket: Current bucket
```
Comment (fhieber) [inline sockeye/data_io.py]: missing period

File: sockeye/train.py [hunk 4]
```diff
@@ -179,9 +187,14 @@ def main():
                                                lr_scheduler=lr_scheduler,
                                                rnn_forget_bias=args.rnn_forget_bias)
 
-        if args.params:
+        # We may consider loading the params in TrainingModule, for consistency
```
Comment (fhieber) [inline sockeye/train.py]: +1
train.py is getting quite big over time and this is inherently logic that belongs to the training module.
TrainingModel.module is built in the constructor of TrainingModel but only initialized within fit()/_fit(), so we could check for resume_training at that point.

File: sockeye/training.py [hunk 5]
```diff
@@ -33,6 +38,18 @@
 
 logger = logging.getLogger(__name__)
 
+class _TrainingState:
```
Comment (fhieber) [inline sockeye/training.py]: Could you add a minimal docstring to this class to make sphinx happy?

File: sockeye/training.py [hunk 6]
```diff
@@ -184,35 +201,33 @@ def fit(self,
             if monitor_bleu else None
 
         logger.info("Training started.")
-        training_monitor = sockeye.callback.TrainingMonitor(train_iter.batch_size, output_folder,
-                                                            optimized_metric=optimized_metric,
-                                                            use_tensorboard=use_tensorboard,
-                                                            checkpoint_decoder=checkpoint_decoder)
+        self.training_monitor = sockeye.callback.TrainingMonitor(train_iter.batch_size, output_folder,
```
Comment (fhieber) [inline sockeye/training.py]: should we set self.training_monitor = None in the constructor then?

File: sockeye/training.py [hunk 7]
```diff
@@ -14,12 +14,17 @@
 """
 Code for training
 """
+from collections import namedtuple
```
Comment (fhieber) [inline sockeye/training.py]: is this used somewhere?

File: sockeye/training.py [hunk 8]
```diff
@@ -223,16 +238,25 @@ def _fit(self,
...
-        epoch = 0
-        checkpoint = 0
-        updates = 0
-        samples = 0
+
+        training_state_dir = os.path.join(output_folder, C.TRAINING_STATE_DIRNAME)
+        if os.path.exists(training_state_dir):
+            training_state = self.load_checkpoint(training_state_dir, train_iter)
+        else:
+            training_state = _TrainingState(
```
Comment (fhieber) [inline sockeye/training.py]: why set the defaults explicitly if you have them in your minimal TrainingState class?
If we make them explicit here we can just as well use typing.NamedTuple and have type annotations as in AttentionState/DecoderState

File: sockeye/training.py [hunk 9]
```diff
@@ -275,16 +299,31 @@ def _fit(self,
...
+
+                if training_state.num_not_improved == max_num_not_improved:
+                    logger.info("Model has not improved for %d checkpoints. Stopping fit.", training_state.num_not_improved)
+                    self.training_monitor.stop_fit_callback()
+                    final_training_state_dirname = os.path.join(output_folder, C.TRAINING_STATE_DIRNAME)
+                    if os.path.exists(final_training_state_dirname):
+                        shutil.rmtree(final_training_state_dirname)
                     break
+                else:
+                    self._checkpoint(training_state, output_folder, train_iter)
```
Comment (fhieber) [inline sockeye/training.py]: moving the checkpoint call to the end of the logic here skips the checkpointing when num_not_improved == max_num_not_improved. Is this what we want ?

File: sockeye/training.py [hunk 10]
```diff
@@ -275,16 +299,31 @@ def _fit(self,
...
 
-    def _evaluate(self, checkpoint, val_iter, val_metric, training_monitor):
+    def _save_params(self, output_folder: str, checkpoint: int):
+        """
+        sync aux params across devices
```
Comment (fhieber) [inline sockeye/training.py]: This method doesn't really sync aux params across devices, the module.get_params() call does.

Since this is an internal method we dont need to have a docstring necessarily, but if we have one, it should describe what the method overall does (saving the current module parameters to a file)

File: sockeye/training.py [hunk 11]
```diff
@@ -296,16 +335,101 @@ def _evaluate(self, checkpoint, val_iter, val_metric, training_monitor):
...
+
+        # Optimizer state (from mxnet)
+        opt_state_fname = os.path.join(training_state_dirname, C.MODULE_OPT_STATE_NAME)
+        if self.bucketing:
+            # This is a bit hacky, as BucketingModule does not provide a
+            # save_optimizer_states call. We take the current active module and
+            # save its state. This should work, as the individual modules in
+            # BucketingModule share the same optimizer through
+            # borrow_optimizer.
+            self.module._curr_module.save_optimizer_states(opt_state_fname)
```
Comment (fhieber) [inline sockeye/training.py]: exposing this should be an easy addition to MXNet's BucketingModule

File: sockeye/training.py [hunk 12]
```diff
@@ -296,16 +335,101 @@ def _evaluate(self, checkpoint, val_iter, val_metric, training_monitor):
...
+        # loss if the process is aborted during deletion (which will be slower
+        # than directory renaming)
+        delete_training_state_dirname = os.path.join(output_folder, C.TRAINING_STATE_TEMP_DELETENAME)
+        if os.path.exists(final_training_state_dirname):
+            os.rename(final_training_state_dirname, delete_training_state_dirname)
+        os.rename(training_state_dirname, final_training_state_dirname)
+        if os.path.exists(delete_training_state_dirname):
+            shutil.rmtree(delete_training_state_dirname)
+
+    def save_state(self, training_state: _TrainingState, fname: str):
```
Comment (fhieber) [inline sockeye/training.py]: missing docstring

File: sockeye/training.py [hunk 13]
```diff
@@ -296,16 +335,101 @@ def _evaluate(self, checkpoint, val_iter, val_metric, training_monitor):
...
+        with open(fname, "wb") as fp:
+            pickle.dump(training_state, fp)
+
+    def load_state(self, fname):
```
Comment (fhieber) [inline sockeye/training.py]: missing docstring (should also include return value & type)

File: test/test_checkpoint.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+# permissions and limitations under the License.
+
+from math import isclose
+import mxnet as mx
+import random
+import tempfile
+
+import sockeye.data_io
+
+def generate_random_sentence(vocab_size, max_len):
```
Comment (fhieber) [inline test/test_checkpoint.py]: this could go into test_utils.py, might be useful in the future

File: test/test_checkpoint.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    length = random.randint(1, max_len)
+    sentence = []
+    for _ in range(length):
```
Comment (fhieber) [inline test/test_checkpoint.py]: Add a comment why +2
Also:
return [random.randint(3, vocab_size + 2) for _ in range(length)]

File: test/test_checkpoint.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        sentence.append(random.randint(3, vocab_size + 2))
+    return sentence
+
+def create_ParallelBucketSentenceIter(source_sentences, target_sentences, max_len):
```
Comment (fhieber) [inline test/test_checkpoint.py]: does pylint accept these? otherwise maybe create_parallel_bucket_sentence_iter

PEP8: two newlines between functions

File: test/test_checkpoint.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (fhieber) [inline test/test_checkpoint.py]: these tests could go into test_data_io.py

File: test/test_checkpoint.py [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+                                                         eos, pad, unk)
+    return bucket_iterator
+
+def data_batches_equal(db1, db2):
+    # We just compare the data, should probably be enough
+    equal = True
+    for data1, data2 in zip(db1.data, db2.data):
+        equal = equal \
+                and (data1.shape == data2.shape) \
+                and isclose(mx.nd.prod(data1.reshape((-1,)) == data2.reshape((-1,))).asnumpy()[0], 1.0)
```
Comment (fhieber) [inline test/test_checkpoint.py]: there is an np.isclose as well

File: test/test_checkpoint.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    # Simulate some iterations
+    ori_iterator.next()
+    ori_iterator.next()
+    ori_iterator.next()
+    ori_iterator.next()
+    expected_output = ori_iterator.next()
+    # expected_output because the user is expected to call next() after loading
+
+    # Save the state to disk
+    tmp_file = tempfile.NamedTemporaryFile()
```
Comment (fhieber) [inline test/test_checkpoint.py]: did you make sure this is deleted afterwards?
You can also use a with block (see test_vocab.py)

File: sockeye/train.py (hunk 4 above)
Comment (davvil) [inline sockeye/train.py]: Completely agree, but will do this in another PR.

File: sockeye/training.py (hunk 7 above)
Comment (davvil) [inline sockeye/training.py]: Not anymore. _TrainingState was initially a namedtuple, but I updated to a class in order to be able to update the values.

File: sockeye/training.py (hunk 6 above)
Comment (davvil) [inline sockeye/training.py]: Yep, will be cleaner.

File: sockeye/training.py (hunk 8 above)
Comment (davvil) [inline sockeye/training.py]: NamedTuples are problematic here because they are inmutable. Agreed that we do not need the defaults in two places.

File: sockeye/training.py (hunk 9 above)
Comment (davvil) [inline sockeye/training.py]: Yes, if the train is finished we delete the state stored by the _checkpoint function.

File: sockeye/training.py (hunk 10 above)
Comment (davvil) [inline sockeye/training.py]: Right, leftover from a previous iteration.

File: test/test_checkpoint.py (hunk 17 above)
Comment (davvil) [inline test/test_checkpoint.py]: Was thinking about it, put them in a new file in the end to leave the door open to additional tests for checkpointing.

File: test/test_checkpoint.py (hunk 14 above)
Comment (davvil) [inline test/test_checkpoint.py]: Will move.

File: test/test_checkpoint.py (hunk 16 above)
Comment (davvil) [inline test/test_checkpoint.py]: Pylint complains, but I found it clearer because of the class name. I am open to changing it, though.

File: test/test_checkpoint.py (hunk 18 above)
Comment (davvil) [inline test/test_checkpoint.py]: Yes, but it converts again into "float booleans" (i.e. 0.0 for false, 1.0 for true). (Actually it should be safe to compare against 1.0)

File: test/test_checkpoint.py (hunk 19 above)
Comment (davvil) [inline test/test_checkpoint.py]: Yep. From the docs: "It will be destroyed as soon as it is closed (including an implicit close when the object is garbage collected)."

File: sockeye/train.py (hunk 4 above)
Comment (fhieber) [inline sockeye/train.py]: makes sense

File: sockeye/training.py (hunk 7 above)
Comment (fhieber) [inline sockeye/training.py]: oh yes, I missed this. Simple class is fine then

File: test/test_checkpoint.py (hunk 18 above)
Comment (fhieber) [inline test/test_checkpoint.py]: ok fine

File: sockeye/train.py [hunk 20]
```diff
@@ -56,6 +58,12 @@ def _build_or_load_vocab(existing_vocab_path: Optional[str], data_path: str, num
         vocabulary = sockeye.vocab.vocab_from_json(existing_vocab_path)
     return vocabulary
 
+def _dict_difference(dict1: Dict, dict2: Dict):
```
Comment (fhieber) [inline sockeye/train.py]: how about:
diff = set(dict1.keys()) - set(dict2.keys())

maybe you don't need a function then anymore

File: sockeye/train.py [hunk 21]
```diff
@@ -79,14 +87,41 @@ def main():
...
         "Must optimize either BLEU or one of tracked metrics (--metrics)"
 
+    # Checking status of output folder, resumption, etc.
     output_folder = os.path.abspath(args.output)
-    if not os.path.exists(output_folder):
+    resume_training = False
+    training_state_dir = os.path.join(output_folder, C.TRAINING_STATE_DIRNAME)
+    if os.path.exists(output_folder):
+        if args.overwrite_output:
+            shutil.rmtree(output_folder)
```
Comment (fhieber) [inline sockeye/train.py]: add a logging message

File: sockeye/train.py [hunk 22]
```diff
@@ -79,14 +87,41 @@ def main():
...
+            os.makedirs(output_folder)
+        elif os.path.exists(training_state_dir):
+            with open(os.path.join(output_folder, C.ARGS_STATE_NAME), "r") as fp:
+                old_args = json.load(fp)
+            arg_diffs = _dict_difference(vars(args), old_args) | _dict_difference(old_args, vars(args))
+            # Remove args that may differ without affecting the training.
+            # TODO: should we include device-ids and use-cpu?
```
Comment (fhieber) [inline sockeye/train.py]: lets not do device-ids since using fewer/more devices affects convergence speed.

File: sockeye/train.py [hunk 23]
```diff
@@ -79,14 +87,41 @@ def main():
...
+            # Are there more?
+            arg_diffs -= set(["overwrite_output", "use-tensorboard", "quiet", "align_plot_prefix", "sure_align_threshold"])
```
Comment (fhieber) [inline sockeye/train.py]: lets define this set in constants.py

File: sockeye/train.py [hunk 24]
```diff
@@ -79,14 +87,41 @@ def main():
...
+            if not arg_diffs:
+                resume_training = True
+            else:
+                # We do not have the logger yet
```
Comment (fhieber) [inline sockeye/train.py]: I see, 

can we create a logger two times?
once before this block without a path argument, and then with path as before?

File: sockeye/train.py (hunk 20 above)
Comment (davvil) [inline sockeye/train.py]: This does not work because we want to compare the values. In fact the keys normally won't differ as they will be set with default values. I put the two ways mainly for "new versions", when arguments have been added or removed.

**Comment by fhieber** [comment]
Thanks a lot for iterating!
